### PR TITLE
#1935: Stunning and Cuffing

### PR DESCRIPTION
**Pull request by JaytleBee**
### Purpose
Implement #1935

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

### Notes:
Still needs some work (apparently it can't automatically merge? I'll need to look into that!). This is to be seen as more of a progress report.

### CODE REVIEW

**Comment by JaytleBee** [comment]
For some reason it tells me there are conflicting files (even though this is merged with the current version of develop). Can someone with more git knowledge have a look at this?

**Comment by Aranclanos** [comment]
Conflicting files
UnityProject/Assets/Scenes/OutpostStation.unity
UnityProject/Assets/Scripts/Weapons/Gun.cs
UnityProject/Assets/Scripts/Weapons/WeaponNetworkActions.cs

**Comment by JaytleBee** [comment]
thanks aranclanos, I can see that. can you maybe see that in more detail or something? github only tells me that those files conflict (even though they shouldn't, I merged with the remote, it should be able to fast-forward)

EDIT: Nevermind, git extensions was being weird. My bad

File: UnityProject/Assets/Scripts/Input System/Triggers/MeleeTrigger.cs.orig [hunk 1]
```diff
@@ -0,0 +1,100 @@
+﻿using System;
```
Comment (Jibbajabbafic) [inline UnityProject/Assets/Scripts/Input System/Triggers/MeleeTrigger.cs.orig]: Is this file needed anymore?

I'd remove it and the meta file if not

**Comment by JaytleBee** [comment]
Alright, I did some more work and it *should* work now. Please test the changes and merge them when you have the time. Git is a bit difficult to work with sometimes...

File: UnityProject/Assets/Scripts/Items/Others/Restraint.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	// TODO: Add time it takes to resist out of handcuffs
+
+	/// <summary>
+	/// Sound to be played when applying restraints
+	/// </summary>
+	public string sound = "Handcuffs";
+
+	protected override bool WillInteract(HandApply interaction, NetworkSide side)
+	{
+		PlayerMove targetPM = interaction.TargetObject?.GetComponent<PlayerMove>();
```
Comment (Jibbajabbafic) [inline UnityProject/Assets/Scripts/Items/Others/Restraint.cs:33]: This will handle all the default interact checks (is player in crit state, in range... etc)
```suggestion
		if (!base.WillInteract(interaction, side)) return false;
		
		PlayerMove targetPM = interaction.TargetObject?.GetComponent<PlayerMove>();
```

File: UnityProject/Assets/Scripts/Player/PlayerMove.cs [hunk 3]
```diff
@@ -317,47 +323,107 @@ public void Unrestrain()
...
+	[Server]
+	public void Uncuff()
+	{
+		cuffed = false;
+
+		pna.DropItem("handcuffs");
+	}
+
+	/// <summary>
+	/// Called when a player
```
Comment (Jibbajabbafic) [inline UnityProject/Assets/Scripts/Player/PlayerMove.cs]: Called when a player...

**Comment by JaytleBee** [comment]
Known bug: When someone is stunned they still collide as if they were standing. I think that's outside of the scope of this issue though and should be addressed separately.

File: UnityProject/Assets/Scripts/Weapons/Melee/MeleeStun.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+/// <summary>
+/// Adding this to a weapon stuns the target on hit
+/// If the weapon has the StunBaton behaviour it only stuns when the baton is active
+/// </summary>
+public class MeleeStun : Interactable<HandApply>
+{
+	/// <summary>
+	/// How long to stun for (in seconds)
+	/// </summary>
+	public float stunTime;
```
Comment (Jibbajabbafic) [inline UnityProject/Assets/Scripts/Weapons/Melee/MeleeStun.cs]: If this is public to set it in the editor I'd suggest doing this instead to make it more obvious:
```suggestion
	[SerializeField]
	private float stunTime;
```

File: UnityProject/Assets/Scripts/Weapons/Melee/MeleeStun.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	private StunBaton stunBaton;
+
+	public void Start()
+	{
+		stunBaton = GetComponent<StunBaton>();
+	}
+
+	protected override bool WillInteract(HandApply interaction, NetworkSide side)
+	{
+		return interaction.UsedObject == gameObject
```
Comment (Jibbajabbafic) [inline UnityProject/Assets/Scripts/Weapons/Melee/MeleeStun.cs:35]:
```suggestion
		if (!base.WillInteract(interaction, side)) return false;

		return interaction.UsedObject == gameObject
```

File: UnityProject/Assets/Scripts/Weapons/Melee/StunBaton.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+[RequireComponent(typeof(Pickupable))]
+[RequireComponent(typeof(StunBatonActivate))]
+public class StunBaton : NetworkBehaviour
+{
+	public SpriteRenderer spriteRenderer;
+
+	/// <summary>
+	/// Sound played when turning this baton on/off
+	/// </summary>
+	public string soundToggle;
```
Comment (Jibbajabbafic) [inline UnityProject/Assets/Scripts/Weapons/Melee/StunBaton.cs]: All these don't need to be public
```suggestion
	[SerializeField]
	private string soundToggle;
```

File: UnityProject/Assets/Scripts/Weapons/Melee/StunBatonActivate.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Collections;
```
Comment (Jibbajabbafic) [inline UnityProject/Assets/Scripts/Weapons/Melee/StunBatonActivate.cs]: I think I'd prefer it if this was merged with `StunBaton.cs` so it's all self-contained, but it's fine

**Review by Jibbajabbafic — COMMENTED** [review]
Code looks pretty good.

I have some small suggestions

File: UnityProject/Assets/Scripts/Input System/Triggers/MeleeTrigger.cs.orig (hunk 1 above)
Comment (JaytleBee) [reply UnityProject/Assets/Scripts/Input System/Triggers/MeleeTrigger.cs.orig]: My bad, that one must have slipped through

File: UnityProject/Assets/Scripts/Weapons/Melee/StunBatonActivate.cs (hunk 7 above)
Comment (JaytleBee) [reply UnityProject/Assets/Scripts/Weapons/Melee/StunBatonActivate.cs]: I actually did it like that but StunBaton has to be a network behaviour (because it uses a syncvar) and StunBatonActivate is an Interaction. There's probably a way to get it all in one file but I don't think it's all that important

Comment (Jibbajabbafic) [reply UnityProject/Assets/Scripts/Weapons/Melee/StunBatonActivate.cs]: You can use `NB(InteractionType)Interactable`, e.g. `NBHandActivateInteractable` 😄

**Comment by Jibbajabbafic** [comment]
Some things we can make issues for:
- Cuffs shouldn't stop movement, just hand interactions
- Doesn't update UI with cuffed sprites:

- No status effect in the corner when cuffed, like for buckling

- No status effect for being stunned either (we should implement a generic status effect system)

- ~~Doesn't make player sprite cuffed either~~
- ~~Handcuffs don't get synced properly after uncuffing (invisible for clients)~~

**Comment by JaytleBee** [comment]
> Handcuffs don't get synced properly after uncuffing (invisible for clients)

Can you check that? This works on my end

**Comment by JaytleBee** [comment]
image
I've tested it with a headless server and two clients and handcuffs show up when you're cuffed and uncuff correctly. This is strange!

**Comment by Jibbajabbafic** [comment]
Small fixes todo:
- Cuffing sound should happen at start of action
- Stun baton sprite scale should be 1

**Comment by Jibbajabbafic** [comment]
Another small thing:
- ~~Uncuffing should have a sound too~~

**Comment by JaytleBee** [comment]
> 
> 
> Another small thing:
> 
>     * Uncuffing should have a sound too

For the benefit of future readers: We have found that it shouldn't.

**Comment by Jibbajabbafic** [comment]
1 more small tweak:
- Toggling the baton should make a sound
